File: distinct-prime-factors-of-product-of-array.py
class Solution:
    def factors(self, n):
        i = 2
        factors = set()
        while i * i <= n:
            if n % i:
                i += 1
            else:
                while n % i == 0:
                    n //= i
                factors.add(i)
        if n > 1:
            factors.add(n)
        return factors
    
    def distinctPrimeFactors(self, nums: List[int]) -> int:
        res = set()
        for el in nums:
            res.update(self.factors(el))
        return len(res)

File: distinct-prime-factors-of-product-of-array.cpp
#include <bitset>
#include <vector>

class Solution {
public:
    Solution() {
        f.resize(1001, std::bitset<1001>(0));
        calculateF();
    }

    int distinctPrimeFactors(std::vector<int>& nums) {
        std::bitset<1001> res;
        for (const auto& el : nums) {
            res |= f[el];
        }
        return res.count();
    }

private:
    std::vector<std::bitset<1001>> f;

    void calculateF() {
        std::bitset<1001> mask(2);

        for (int i = 2; i <= 1000; ++i) {
            mask <<= 1;
            int j = 1;
            if (f[i].any()) {
                continue;
            }
            while (i * j <= 1000) {
                f[i * j] |= mask;
                j += 1;
            }
        }
    }
};